compiler/src/iree/compiler/InputConversion/Common/QuantizedMatmulToMatmul.cpp

```diff
@@ -23,16 +23,169 @@ namespace mlir::iree_compiler {
 
 namespace {
 
-bool isConstantZero(Value val) {
+static bool isConstantZero(Value val) {
   auto constIntOp = val.getDefiningOp<arith::ConstantIntOp>();
   return constIntOp && constIntOp.value() == 0;
 }
 
 // Pattern lowering quantized_matmul to matmul.
+// This also targets the batch_matmul op.
 // Always succeeds.
 //
 // This is implementing the math explained in Section 2.3 of
 // https://arxiv.org/abs/1712.05877.
+static Value computeQuantizedMatmul(Location loc, ValueRange &inputs,
```

Comment:
Also regarding the function's interface here:
1. The name, `computeQuantizedMatmul`, is a bit hard to understand, I guess the verb "compute" is rarely used in this sense. I would rather name this `lowerQuantizedMatmul` and maybe make this return a `linalg::LinalgOp` rather than a `Value` (returning the operation itself, not merely its return value, though that's ultimately equivalent) or, if the `replaceOp` was moved into this func, the name `rewriteQuantizedMatmul` would become possible (and then the function could just return `void` or a status).
2. I wonder if the following might be just a simpler interface:

```c++
void rewriteQuantizedMatmul(linalg::LinalgOp op, PatternRewriter& rewriter)
```

Feel free to change the return type to return a status if needed; and the idea is that any linalg Op, such as a QuantizedMatmulOp or a QuantizedBatchMatmulOp, is in particular a linalg::LinalgOp, so you can just pass them like that (and you don't even need a c++ template parameter for that) and then inside the function you can do things like `if(isa<linalg::QuantizedBatchMatmulOp>(op)) {... } `.

Reply:
I see. We didn't need a new function at all. Everything can be embedded inside the previous rewrite. Thanks to templates. Also, thanks for the detailed review. My C++ skills are getting rusty these days. I need to work hard. 😄

Reply:
C++ templates are a powerful tool, and we do use it sometimes -- but it does have costs (harder to read C++ source, inflated binary) so we like to avoid it when there isn't a strong need for it.

Here, I believe that C++ templates are unnecessary: this Pattern could simply be on `linalg::LinalgOp` and inside it you could do `if(isa<linalg::QuantizedBatchMatmulOp>(op)) {... } `,  doing *at runtime* the logic that C++ templates allowed moving to *compile time*.

Can you give it a try?

Reply:
Note: a nuance will be that, now that your pattern is on `linalg::LinalgOp`, it will need to `return failure();` when the op is not a QuantizedMatmulOp or a QuantizedBatchMatmulOp.

Reply:
yes!

Reply:
This throws `:364:23: error: no member named 'getOperationName' in 'mlir::linalg::LinalgOp'`.

Reply:
Oh I see. I grepped,

```
(bjacob-ukernel-cconv) ~/iree git grep -n 'RewritePattern<linalg::LinalgOp>'
compiler/src/iree/compiler/Codegen/Common/ConvertToDestinationPassingStylePass.cpp:511:    : public OpInterfaceRewritePattern<linalg::LinalgOp> {
compiler/src/iree/compiler/Codegen/Common/ConvertToDestinationPassingStylePass.cpp:512:  using OpInterfaceRewritePattern<linalg::LinalgOp>::OpInterfaceRewritePattern;
compiler/src/iree/compiler/GlobalOptimization/DetachElementwiseFromNamedOps.cpp:33:    : public OpInterfaceRewritePattern<linalg::LinalgOp> {
compiler/src/iree/compiler/GlobalOptimization/PropagateLinalgTranspose.cpp:388:    : public OpInterfaceRewritePattern<linalg::LinalgOp> {
compiler/src/iree/compiler/GlobalOptimization/PropagateLinalgTranspose.cpp:390:  using OpInterfaceRewritePattern<linalg::LinalgOp>::OpInterfaceRewritePattern;
compiler/src/iree/compiler/GlobalOptimization/SetEncoding.cpp:292:    : public OpInterfaceRewritePattern<linalg::LinalgOp> {
compiler/src/iree/compiler/GlobalOptimization/SetEncoding.cpp:293:  using OpInterfaceRewritePattern<linalg::LinalgOp>::OpInterfaceRewritePattern;
compiler/src/iree/compiler/Preprocessing/Common/ConvertConvToChannelsLast.cpp:461:struct ConvertLinalgConvOp : OpInterfaceRewritePattern<linalg::LinalgOp> {
compiler/src/iree/compiler/Preprocessing/Common/ConvertConvToChannelsLast.cpp:462:  using OpInterfaceRewritePattern<linalg::LinalgOp>::OpInterfaceRewritePattern;
compiler/src/iree/compiler/Preprocessing/Common/ConvertConvToChannelsLast.cpp:465:      : OpInterfaceRewritePattern<linalg::LinalgOp>(context, benefit),
compiler/src/iree/compiler/Preprocessing/Common/PadLinalgOps.cpp:22:class PadMatmulOp : public OpInterfaceRewritePattern<linalg::LinalgOp> {
```

and so what this says is that you need `OpInterfaceRewritePattern`, not `OpRewritePattern` . Sorry about that.

Reply:
Done, thanks.